### DISCUSSION TITLE
How to use the JSBSim plugin with Unreal Engine?

### DISCUSSION THREAD

**Discussion by JoinKomorebi** [JSBSim-Team/jsbsim#664] [Q&A] [answered]
I recently downloaded the source code of JSBSim on Git, and I want to use this plugin in Unreal Engine. I run the file JSBSimForUnreal.sln first, and then I want to open the project under UnrealEngine, and I encountered the following problem
![9f13406a211928059680535e0eaa8eb](https://user-images.githubusercontent.com/104044166/174982350-735a6e77-3073-4596-8e40-112b5708a99b.png)
![d1d1fb27103a8cae9197b6a3937b330](https://user-images.githubusercontent.com/104044166/174982433-eed24a91-2c94-49d6-b9fd-2fe31f91b347.png)
![6775a4858aa3a7bd91fabab6c2b9b27](https://user-images.githubusercontent.com/104044166/174982493-def3df26-4942-44cc-857a-75e5623063c4.png)
s
I also tried to use Visual Studio 2019 to compile, but it was still unsuccessful. Do you guys know the reason?Hope you guys can tell me why.thank you very much

**Comment by bcoconni** [comment]
Did you follow the steps described in the [plugin README](https://github.com/JSBSim-Team/jsbsim/blob/master/UnrealEngine/README.md) ?

**Reply by JoinKomorebi** [reply]
I have tried it all. After I can't open the Unreal project, I generate the solution, and it will report an error when I run it.
![image](https://user-images.githubusercontent.com/104044166/175229592-ffd4e8a5-d921-4ee4-bf35-6ee08d9a5606.png)

**Reply by bcoconni** [reply]
May be @AlbanBERGERET-Epic or @gallonmate could help you ?

**Reply by JoinKomorebi** [reply]
Thanks for your help, I have solved the problem

**Comment by gallonmate** [comment]
When you ran JSBSimForUnreal.sln, was the build successful? The error is saying jsbsim.dll is not found. Check your folder to see what's there. If they are not there, then I'd delete everything and start fresh. This is example of the files you should see in \UnrealEngine\Plugins\JSBSimFlightDynamicsModel\Source\ThirdParty\JSBSim\Lib
![image](https://user-images.githubusercontent.com/28702753/175625909-c81f49d0-31ff-4d1f-86e8-531f0725907b.png)

**Reply by JoinKomorebi** [reply]
I found that after I run JSBSimForUnreal.sln the generated file directory is like this。So the log prompt in vs is that the jsbsim.dll file cannot be found.I don't know what's causing it.
![512d19cb9b4e98529f0e3eb8e14afe1](https://user-images.githubusercontent.com/104044166/176335538-813d06e9-0365-4488-ac80-37e3129082f2.png)
![4e3d6e7b018e357f565e3a9f0870fa4](https://user-images.githubusercontent.com/104044166/176335573-812563c5-d04e-4924-8923-71b4ade66bcb.png)
s

**Reply by AlbanBERGERET-Epic** [reply]
The produced files are in libdebug folder, meaning that you built jsbsimforunreal in debug mode. 
You need to build it in Release mode. 

The libdebug version is in case you want to use jsbsim debug libs in unreal, and for that you need to build the uereferenceApp in DebugGame. In most use case, this is not needed.

=› build jsbsimforunreal in release, and uereferenceApp in development modes.

**Comment by AlbanBERGERET-Epic** [comment, marked as answer]
The error message is pretty explicit. You need to build jsbsim.dll first. When opened in Visual Studio, you should select and build x64/Release. 
Do that first, and in case it fails, share your build log here...

**Reply by JoinKomorebi** [reply]
Thank you for your answer, I have solved the problem, I have not set to Release version

**Reply by ahmadmasoud** [reply]
![image](https://user-images.githubusercontent.com/47717125/177371865-8fd6d6cc-2f03-4068-8b3d-9c960cba80c0.png)

**Reply by AlbanBERGERET-Epic** [reply]
@ahmadmasoud 
My guess is that the copy script fails because there is a space in your path (C:\users\Ahmad Massoud\Documents...) 

It could probably be solved by adding a couple of "" around variables in the post-build event script, but generally, having spaces in paths is not recommended. Moreover, long paths (>255 chars) can sometimes be an issue for some scripts or libs (not strictly ue-related). 

My recommendation would be to locate your repository somewhere else, without spaces, and with shorter paths. 
Like "D:\jsbsim-master" for example.

**Comment by gallonmate** [comment]
battlegear2099, I can't see your message on github right now. So I assume
you solved and deleted your comment? If not solved, then you can just find
a github commit from when the UE5 plugin was last updated or try a commit
from a week ago. They are currently worked on some string conversion
updates for JSBSim, which appear are giving you that error and failed build.

**Comment by seanmcleod** [comment]
battlgear2099, @gallonmate I forgot that there is a separate Visual Studio project file for the Unreal stuff, so I only updated standard JSBSim Visual Studio project file. In the meantime @bcoconni has pushed a commit to update the Unreal Visual Studio project file - https://github.com/JSBSim-Team/jsbsim/commit/a8c91ec889288a8e5ae127e8eca9bd359e035420

**Comment by seanmcleod** [comment]
> And there is a error about something missing that jbsim needs.

@techjunkie2099 that's the whole point about the update to the Visual Studio project file that I mentioned above, i.e. a reference has been added for the new source file so that you won't get an error from the linker. So just pull the repo to get the updated Visual Studio project file.
